Align the operator<< for Argument with FunctionSchema parser

houseroad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23208 expose parse_schema and __eq__ function to python and add round trip tests
* #23207 Expose an API to iterate all the registered operators
* #23206 Align AliasInfo's operator<< with FunctionSchema
* #23205 Support IntList in Dict's shalloEquals
* #23204 Support variadic returns in Schema's operator<<
* **#23203 Align the operator<< for Argument with FunctionSchema parser**
* #23202 Fix the operator== for Argument
* #23201 Align String str() with the format in FunctionSchema
* #23200 Align Dict str() with the format in FunctionSchema
* #23199 Support parsing variadic return schema

Align the Argument's operator<< with parser,
additional support:
1) List size
2) real default value
3) Alias information

Differential Revision: [D16433188](https://our.internmc.facebook.com/intern/diff/D16433188/)

houseroad committed:
Align the operator<< for Argument with FunctionSchema parser

Align the Argument's operator<< with parser,
additional support:
1) List size
2) real default value
3) Alias information

Differential Revision: [D16433188](https://our.internmc.facebook.com/intern/diff/D16433188/)

houseroad committed:
Update on "Align the operator<< for Argument with FunctionSchema parser"


Align the Argument's operator<< with parser,
additional support:
1) List size
2) real default value
3) Alias information

Differential Revision: [D16433188](https://our.internmc.facebook.com/intern/diff/D16433188/)

houseroad committed:
Update on "Align the operator<< for Argument with FunctionSchema parser"


Align the Argument's operator<< with parser,
additional support:
1) List size
2) real default value
3) Alias information

Differential Revision: [D16433188](https://our.internmc.facebook.com/intern/diff/D16433188/)

houseroad committed:
Update on "Align the operator<< for Argument with FunctionSchema parser"


Align the Argument's operator<< with parser,
additional support:
1) List size
2) real default value
3) Alias information

Differential Revision: [D16433188](https://our.internmc.facebook.com/intern/diff/D16433188/)

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -243,7 +243,49 @@ inline bool operator!=(const FunctionSchema& lhs, const FunctionSchema& rhs) {
 
 // for debugging, make sure we can describe the call site
```
I think it's not for debugging any more

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -243,7 +243,49 @@ inline bool operator!=(const FunctionSchema& lhs, const FunctionSchema& rhs) {
+  // so we always use Type(alias)? format
+  std::stringstream oss;
+  if (arg.type()->isSubclass(TypeKind::ListType) && arg.N()) {
+    oss << std::static_pointer_cast<const ListType>(arg.type())->
```
there's .cast<>() method

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -243,7 +243,49 @@ inline bool operator!=(const FunctionSchema& lhs, const FunctionSchema& rhs) {
+  }
+  if (arg.default_value()) {
+    out << "=";
+    //out << expand_escapes(arg.default_value().value().toStringRef());
```
remove?

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -243,7 +243,49 @@ inline bool operator!=(const FunctionSchema& lhs, const FunctionSchema& rhs) {
+    //out << expand_escapes(arg.default_value().value().toStringRef());
+    if (arg.type()->kind() == c10::TypeKind::StringType) {
+        out << "\'";
+        std::ios_base::fmtflags flags(out.flags());
```
nit: I'd add TODO here to "prettify it". Also I wonder whether we should just fix IValue's operator<< to print string literal directly

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -243,7 +243,49 @@ inline bool operator!=(const FunctionSchema& lhs, const FunctionSchema& rhs) {
+    out << "=";
+    //out << expand_escapes(arg.default_value().value().toStringRef());
+    if (arg.type()->kind() == c10::TypeKind::StringType) {
+        out << "\'";
```
shouldn't it be " not ' ? or which syntax do we use?

houseroad replied:
both work in this case, check https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/lexer.h#L200

Also, ''' and """ will work too.

I chose ' since i saw examples here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/register_string_ops.cpp#L455

Just to be consistent.

houseroad committed:
Update on "Align the operator<< for Argument with FunctionSchema parser"


Align the Argument's operator<< with parser,
additional support:
1) List size
2) real default value
3) Alias information

Differential Revision: [D16433188](https://our.internmc.facebook.com/intern/diff/D16433188/)
